### PR TITLE
fix: handle undefined mediaDevices

### PR DESCRIPTION
**Pull request by OBe95**
# Description

Fix: [useMediaDevices(): TypeError: Cannot read property 'addEventListener' of undefined](https://github.com/streamich/react-use/issues/820)

## Type of change

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] **Breaking change** _(fix or feature that would cause existing functionality to not work as before)_

# Checklist
- [x] Read the [Contributing Guide](https://github.com/streamich/react-use/blob/master/CONTRIBUTING.md)
- [x] Perform a code self-review
- [-] Comment the code, particularly in hard-to-understand areas
- [-] Add documentation
- [-] Add hook's story at Storybook
- [-] Cover changes with tests
- [x] Ensure the test suite passes (`yarn test`)
- [-] Provide 100% tests coverage
- [x] Make sure code lints (`yarn lint`). Fix it with `yarn lint:fix` in case of failure.
- [-] Make sure types are fine (`yarn lint:types`).

### CODE REVIEW

File: src/util.ts [hunk 1]
```diff
@@ -1,5 +1,23 @@
...
+  } catch (error) {
+    handleError(error);
+  }
+};
 
-export const off = (obj: any, ...args: any[]) => obj.removeEventListener(...args);
+export const off = (obj: any, ...args: any[]) => {
+  try {
+    obj.removeEventListener(...args);
+  } catch (error) {
```
Comment (streamich) [inline src/util.ts]: I'm not sure we should be swallowing these errors. Also, these utilities are use by other hooks, too, not only by the `useMediaDevices` hook, if we want to swallow the errors, would be better to do it in the `useMediaDevice` hook directly.

Comment (OBe95) [reply src/util.ts]: Well, I don't see why `useMediaDevices` should have a different behavior, especially, if the source of the issue is the same.
As we could have similar issues for another `addEventListener` or `removeEventListener`, I thought we'd rather handle it globally and swallow the errors in PROD env (unless I misunderstood the issue) ?

Comment (streamich) [reply src/util.ts]: I don't think we should be swallowing these errors. You are swallowing the errors in all environments. In PROD you are not logging.

Comment (OBe95) [reply src/util.ts]: Ah! I see you point (I interpreted swallowing as not logging).
I clearly misunderstood what was expected in the issue, how would you approach it ?
Doing it only for the `useMediaDevices` hook or just checking if the `obj` is defined before manipulating the event listeners ?

Comment (streamich) [reply src/util.ts]: Yes, checking if object exists before manipulating it sounds good.

Comment (OBe95) [reply src/util.ts]: Done in a new commit (don't know what you are following in the project, you rather update the commit or add new commit for feedback ?)

File: src/util.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 export const isClient = typeof window === 'object';
 
-export const on = (obj: any, ...args: any[]) => obj.addEventListener(...args);
+export const on = (obj: any, ...args: any[]) => obj && obj.addEventListener(...args);
```
Comment (xobotyi) [inline src/util.ts]: Not sure if these functions should be touched at all. With your changes in some cases handler bind will silently fail and evelopment can be hardened. Better to have error in that case.

You should leave only changes in `src/useMediaDevices.ts`, update the PR and ill merge it.

**Comment by xobotyi** [comment]
@streamich maybe make a breaking change and change the returning signature?
I dont see any reason to return object with a single property.

File: src/util.ts (hunk 2 above)
Comment (xobotyi) [reply src/util.ts]: As you forbid to make changes in your fork repo it should be like so:

**`useMediaDevices`:**
```typescript
import { useEffect, useState } from 'react';
import { noop, off, on } from './util';

const useMediaDevices = () => {
  const [state, setState] = useState({});

  useEffect(() => {
    if (!navigator.mediaDevices) return;

    let mounted = true;

    const onChange = () => {
      navigator.mediaDevices
        .enumerateDevices()
        .then(devices => {
          mounted &&
            setState({
              devices: devices.map(({ deviceId, groupId, kind, label }) => ({ deviceId, groupId, kind, label })),
            });
        })
        .catch(noop);
    };

    on(navigator.mediaDevices, 'devicechange', onChange);
    onChange();

    return () => {
      mounted = false;
      off(navigator.mediaDevices, 'devicechange', onChange);
    };
  }, []);

  return state;
};

export default useMediaDevices;
```

**`util`**
```typescript
export const isClient = typeof window === 'object';

export const on = (obj: any, ...args: any[]) => obj.addEventListener(...args);

export const off = (obj: any, ...args: any[]) => obj.removeEventListener(...args);

export const noop = () => {};
```

**Pull request merged**